test/micro/org/openjdk/bench/javax/crypto/full/BenchBase.java

```diff
@@ -0,0 +1,136 @@
+ * single and multi-part testing.
+ *
+ * The IV rotates through a set buffer and creates a new AlgorithmParameterSpec
+ * for each encrypt benchmark operation.
+ */
+
+public abstract class BenchBase extends CryptoBase {
+    // Defined by the test
+    String algorithm;
+    int keyLength = 256;
+
+    // Default data sizes for full tests
+    @Param({"1024", "1500", "4096", "16384"})
+    int dataSize;
+
+    static final int IV_BUFFER_SIZE = 24;
+    public static final int IV_MODULO = 12;
+    public byte[] iv;
+    public int iv_index = 0;
+    int updateLen = 0;
+
+    private Cipher encryptCipher, decryptCipher;
+    byte[] encryptedData, in, out;
+    SecretKeySpec ks;
+    AlgorithmParameterSpec spec;
+
+    abstract AlgorithmParameterSpec getNewSpec();
+
+    // Configure setup with particular test parameters
+    public void init(String algorithm, int keyLength) throws Exception {
+        this.algorithm = algorithm;
+        this.keyLength = keyLength;
+        init();
+    }
+
+    // Configure setup with particular test parameters
+    public void init(String algorithm, int keyLength, int dataSize)
+        throws Exception {
+        this.dataSize = dataSize;
+        init(algorithm, keyLength);
+    }
+
+    // Initalize test setup
+    private void init() throws Exception {
+        setupProvider();
+
+        // Setup key material
+        iv = fillSecureRandom(new byte[IV_BUFFER_SIZE]);
+        spec = getNewSpec();
+        // CC20 doesn't care about the algorithm name on the key, but AES does.
+        ks = new SecretKeySpec(fillSecureRandom(new byte[keyLength / 8]),
+            "AES");
+
+        // Setup Cipher classes
+        encryptCipher = makeCipher(prov, algorithm);
+        encryptCipher.init(Cipher.ENCRYPT_MODE, ks, spec);
+        decryptCipher = makeCipher(prov, algorithm);
+        decryptCipher.init(Cipher.DECRYPT_MODE, ks,
+            encryptCipher.getParameters(). getParameterSpec(spec.getClass()));
+
+        // Setup input/output buffers
+        in = fillRandom(new byte[dataSize]);
+        encryptedData = new byte[encryptCipher.getOutputSize(in.length)];
+        out = new byte[encryptedData.length];
+        encryptCipher.doFinal(in, 0, in.length, encryptedData, 0);
+        updateLen = in.length / 2;
+
+    }
+
+    @Benchmark
+    public void encrypt() throws Exception {
+        encryptCipher.init(Cipher.ENCRYPT_MODE, ks, getNewSpec());
+        encryptCipher.doFinal(in, 0, in.length, out, 0);
+    }
+
+    @Benchmark
+    public void encryptMultiPart() throws Exception {
+        encryptCipher.init(Cipher.ENCRYPT_MODE, ks, getNewSpec());
+        int outOfs = encryptCipher.update(in, 0, updateLen, out, 0);
+        encryptCipher.doFinal(in, updateLen, in.length - updateLen,
+            out, outOfs);
+    }
+
+    @Benchmark
+    public void decrypt() throws Exception {
+        decryptCipher.init(Cipher.DECRYPT_MODE, ks,
+            encryptCipher.getParameters().
```

Comment:
Consider using a different method to obtain parameter spec here; the flamegraph suggests that `getParameters` is responsible for up to 20% of the time spent in these benchmarks

Reply:
Thanks for point that out.  I have decided to leave this as is.  There are optimizations that can be made to `getParameters` that I would like the benchmark to measure.

Reply:
There was no optimization available for `getParameters` so I saved the parameters from the encryption op and it greatly improved the test performance